napari-workshops/notebooks/spot_detection_functions.md

```diff
@@ -0,0 +1,408 @@
+
+In this module, we will implement elements of our previous workflow as functions and then use [`magicgui.magicgui`](https://pyapp-kit.github.io/magicgui/api/magicgui/#magicguimagicgui) decorator on those functions to return us compound widgets that we can use to make exploring the parameters easier in the GUI. For a nice overview of the `magicgui` decorators, see [the official documentation](https://pyapp-kit.github.io/magicgui/decorators/).
+
+Let's get everything set up, based on the previous notebook:
+
+```{code-cell} ipython3
+from skimage import io
+
+nuclei_url = 'https://raw.githubusercontent.com/kevinyamauchi/napari-spot-detection-tutorial/main/data/nuclei_cropped.tif'
+nuclei = io.imread(nuclei_url)
+
+spots_url = 'https://raw.githubusercontent.com/kevinyamauchi/napari-spot-detection-tutorial/main/data/spots_cropped.tif'
+spots = io.imread(spots_url)
+```
+
+```{code-cell} ipython3
+import napari
+from napari.utils import nbscreenshot
+
+# create the napari viewer
+viewer = napari.Viewer()
+
+# add the nuclei image to the viewer
+viewer.add_image(nuclei, colormap = 'I Forest', blending = 'minimum')
+
+# add the spots image to the viewer
+viewer.add_image(spots, colormap = 'I Orange', blending='minimum')
+```
+
+Now let's write a function that takes an array and a `sigma` value and performs the 
+high-pass operation.
+
+```{code-cell} ipython3
+import numpy as np
+from scipy import ndimage as ndi
+
+def gaussian_high_pass(image, sigma):
+    low_pass = ndi.gaussian_filter(image, sigma)
+    high_passed_im = (image - low_pass).clip(0)
+    
+    return high_passed_im
+```
+
+We can test our function, in similar fashion as before:
+
+```{code-cell} ipython3
+high_passed_spots = gaussian_high_pass(spots, 2)
+
+viewer.add_image(high_passed_spots, colormap="I Blue", blending="minimum")
+```
+
+```{code-cell} ipython3
+nbscreenshot(viewer)
+```
+
+## Obtaining a basic widget using the `@magicgui` decorator
+
+Now lets modify the function slightly, by providing type annotations and a docstring, to 
+leverage napari `magicgui` integration.
+
+````{tip}
+A brief note about type hints:  
+Type hints are not enforced at runtime, **but** they can still raise `NameError` exceptions if not defined or imported.
+To avoid that, we are putting the napari types in quotes to make them "forward references", because we have
+not imported them. Alternatively, we could have imported them. A third option is to import:
+
+```python
+from __future__ import annotations 
+```
+
+This would permit us to drop the quotes from the type hints.  
```

Comment:
Can't these be already be dropped in Py 311? (and even 310?)

Reply:
If I'm reading https://peps.python.org/pep-0649/ and https://peps.python.org/pep-0749/ correctly, then this note will remain correct until python 3.14 when the default behavior will be the `from __future__ import annotations` behavior.
Also while this workshop uses 3.11, I think it's useful to have stuff work over the range of napari supported versions.

Reply:
> python 3.14

πthon!